code/modules/atmospherics/FEA_system.dm

```diff
@@ -1,322 +1,280 @@
-		var/list/turf/simulated/possible_borders
-		var/list/turf/simulated/possible_space_borders
-		var/possible_space_length = 0
-
-		while(possible_members.len > 0) //Keep expanding, looking for new members
-			for(var/turf/simulated/test as anything in possible_members)
-				test.length_space_border = 0
-				for(var/direction in cardinal)
-					var/turf/T = get_step(test,direction)
-					if(T && !(T in members) && test.gas_cross(T))
-						if(istype(T,/turf/simulated))
-							if(!T:parent)
-								possible_members += T
-								members += T
-							else
-								LAZYLISTINIT(possible_borders)
-								possible_borders |= test
-						else if(istype(T, /turf/space) && !istype(T, /turf/space/fluid))
-							LAZYLISTINIT(possible_space_borders)
-							possible_space_borders |= test
-							test.length_space_border++
-
-				if(test.length_space_border > 0)
-					possible_space_length += test.length_space_border
-				possible_members -= test
-
-		if(members.len > 1)
-			var/datum/air_group/group = new
-			if(possible_borders && (possible_borders.len > 0))
-				group.borders = possible_borders
-			if(possible_space_borders && (possible_space_borders.len > 0))
-				group.space_borders = possible_space_borders
-				group.length_space_border = possible_space_length
-
-			for(var/turf/simulated/test as anything in members)
-				test.parent = group
-				test.processing = 0
-				active_singletons -= test
-
-				test.dist_to_space = null
-				var/dist
-				for(var/P in possible_space_borders)
-					var/turf/simulated/b = P
-					if (b == test)
-						test.dist_to_space = 1
-						break
-					dist = GET_DIST(b, test)
-					if ((test.dist_to_space == null) || (dist < test.dist_to_space))
-						test.dist_to_space = dist
-
-			// Allow groups to determine if group processing is applicable after FEA setup
-			if(current_cycle) group.group_processing = FALSE
-			group.members = members
-			air_groups += group
-
-			group.update_group_from_tiles() //Initialize air group variables
-			return group
-		else
-			base.processing = 0 //singletons at startup are technically unconnected anyway
-			base.parent = null
-
-			if(base.air && base.air.check_tile_graphic())
-				base.update_visuals(base.air)
-
-		return null
-
-	process()
-		current_cycle++
-
-		process_tiles_to_space()
-		is_busy = TRUE
-
-		if(!explosions.exploding)
-			if(groups_to_rebuild.len > 0)
-				process_rebuild_select_groups()
-			LAGCHECK(LAG_REALTIME)
+	var/turf/space/space_sample //instead of repeatedly using locate() to find space, we should just cache a space tile ok
+
+/// Updates cached space sample if need be.
+// Returns: New space sample.
+/datum/controller/air_system/proc/update_space_sample()
+	if (!space_sample || !(space_sample.turf_flags & CAN_BE_SPACE_SAMPLE))
+		space_sample = locate(/turf/space)
+	return space_sample
+
+/datum/controller/air_system/proc/setup(datum/controller/process/air_system/controller)
+	parent_controller = controller
+
+	#if SKIP_FEA_SETUP == 1
```

Comment:
is `SKIP_FEA_SETUP` a bool btw? If it isn't, you might want to use defines or something. That or `#ifdef`.

Reply:
it reaches into other files so not touching

Reply:
fair